otter/convergence/service.py

```diff
@@ -56,99 +62,363 @@ def all_met(server, current_lb_nodes):
+
+    def update_group_state(group, old_state):
+        return assoc_obj(old_state, active=active)
+
+    return Effect(ModifyGroupState(scaling_group=scaling_group,
+                                   modifier=update_group_state))
+
+
+def execute_convergence(tenant_id, group_id, log,
+                        get_all_convergence_data=get_all_convergence_data):
     """
     Gather data, plan a convergence, save active and pending servers to the
     group state, and then execute the convergence.
 
-    :param IScalingGroup scaling_group: The scaling group object.
-    :param GroupState group_state: The group state.
-    :param launch_config: An otter launch config.
-    :param now: The current time in seconds.
-    :param log: A bound logger.
-    :param get_all_convergence_data: The :func`get_all_convergence_data` to use
+    :param group_id: group id
+    :param log: bound logger
+    :param get_all_convergence_data: like :func`get_all_convergence_data`, used
         for testing.
 
     :return: An Effect of a list containing the individual step results.
+    :raise: :obj:`NoSuchScalingGroupError` if the group doesn't exist.
     """
-    all_data_eff = get_all_convergence_data(scaling_group.uuid)
-
-    def got_all_data((servers, lb_nodes)):
-        active = determine_active(servers, lb_nodes)
-        desired_group_state = get_desired_group_state(
-            scaling_group.uuid, launch_config, desired_capacity)
-        steps = plan(desired_group_state, servers, lb_nodes, now)
-        active = {server.id: server_to_json(server) for server in active}
+    # Huh! It turns out we can parallelize the gathering of data with the
+    # fetching of the scaling group info from cassandra.
+    sg_eff = Effect(GetScalingGroupInfo(tenant_id=tenant_id,
+                                        group_id=group_id))
+    gather_eff = get_all_convergence_data(group_id)
+
+    def got_all_data(((scaling_group, manifest), (servers, lb_nodes))):
+        group_state = manifest['state']
+        launch_config = manifest['launchConfiguration']
+        time_eff = Effect(Func(time.time))
+
+        def got_time(now):
+            desired_group_state = get_desired_group_state(
+                group_id, launch_config, group_state.desired)
+            steps = plan(desired_group_state, servers, lb_nodes, now)
+            active = determine_active(servers, lb_nodes)
+            log.msg('execute-convergence',
+                    servers=servers, lb_nodes=lb_nodes, steps=steps, now=now,
+                    desired=desired_group_state, active=active)
+            eff = _update_active(scaling_group, active)
+            eff = eff.on(lambda _: steps_to_effect(steps))
+            return eff
+        return time_eff.on(got_time)
+
+    return Effect(TenantScope(
+        parallel([sg_eff, gather_eff]).on(
+            success=got_all_data,
+            error=catch(FirstError, lambda fe: six.reraise(*fe[1].exc_info))),
```

Comment:
@manishtomar - thanks, I've added six. It was already being depended upon transitively, but it's good to be explicit.